wiring/src/spark_wiring_logging.cpp

```diff
@@ -0,0 +1,246 @@
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <algorithm>
+#include <cstdio>
+
+#include "spark_wiring_logging.h"
+
+namespace {
+
+using spark::LogHandler;
+
+class LogManager {
+public:
+    void addHandler(LogHandler *handler) {
+        const auto it = std::find(handlers_.cbegin(), handlers_.cend(), handler);
+        if (it == handlers_.end()) {
+            handlers_.push_back(handler);
+            if (handlers_.size() == 1) {
+                log_set_callbacks(logMessage, logWrite, logEnabled, nullptr);
+            }
+        }
+    }
+
+    void removeHandler(LogHandler *handler) {
+        const auto it = std::find(handlers_.begin(), handlers_.end(), handler);
+        if (it != handlers_.end()) {
+            if (handlers_.size() == 1) {
+                log_set_callbacks(nullptr, nullptr, nullptr, nullptr);
+            }
+            handlers_.erase(it);
+        }
+    }
+
+    static LogManager* instance() {
+        static LogManager mgr;
+        return &mgr;
+    }
+
+private:
+    std::vector<LogHandler*> handlers_;
+
+    static void logMessage(const char *msg, int level, const char *category, uint32_t time, const char *file, int line,
+            const char *func, void *reserved) {
+        const LogHandler::SourceInfo info = {file, func, line};
+        const auto &handlers = instance()->handlers_;
+        for (size_t i = 0; i < handlers.size(); ++i) {
+            handlers[i]->message(msg, (LogLevel)level, category, time, info);
+        }
+    }
+
+    static void logWrite(const char *data, size_t size, int level, const char *category, void *reserved) {
+        const auto &handlers = instance()->handlers_;
+        for (size_t i = 0; i < handlers.size(); ++i) {
+            handlers[i]->write(data, size, (LogLevel)level, category);
+        }
+    }
+
+    static int logEnabled(int level, const char *category, void *reserved) {
+        int minLevel = LOG_LEVEL_NONE;
+        const auto &handlers = instance()->handlers_;
+        for (size_t i = 0; i < handlers.size(); ++i) {
+            const int level = handlers[i]->categoryLevel(category);
+            if (level < minLevel) {
+                minLevel = level;
+            }
+        }
+        return (level >= minLevel);
+    }
+};
+
+inline const char* strchrend(const char* s, char c) {
+    const char* result = strchr(s, c);
+    return result ? result : s + strlen(s);
+}
+
+} // namespace
+
+// spark::LogHandler
+struct spark::LogHandler::FilterData {
+    const char *name; // Category name
+    size_t size; // Name length
+    int level; // Logging level (-1 if not specified for this category)
+    std::vector<FilterData> filters; // Subcategories
+
+    FilterData(const char *name, size_t size) :
+            name(name),
+            size(size),
+            level(-1) {
+    }
+};
+
+spark::LogHandler::LogHandler(Stream *stream, LogLevel level, const Filters &filters) :
```

Comment:
I'd really like to see this constructor refactored into several smaller private methods with descriptive names. It's really dense algorithmic code and it's hard to guess what each piece is trying to accomplish.